### PR TITLE
Allow using physical DPI for map canvas

### LINKED ISSUE #44128: Symbology size on HiDPI screens
On HiDPI screens, symbol sizes are not correctly represented.

The screenshot highlights the problem, square size on this paper is 5 mm, but circles that should be the same size are only roughly half of it.

Last time I checked (a couple of years ago), this did not affect low res displays (96dpi).

This leads to QGIS projects which are represented:
- incorrectly
- differently, depending on DPI
- differently [on QField which respects DPI](https://github.com/opengisch/QField/issues/1970)

PXL_20210712_152900156

### PR DESCRIPTION
**Pull request by m-kuhn**
## Description
This introduces a new setting to correctly use physical screen DPI instead of logical DPI.
It is off by default to make sure users don't suddenly get symbols scaled differently than before.

This makes symbols appear equal on different attached devices (hi dpi and low dpi) and on qfield

I would recommend to switch this setting for QGIS 4.

Fixes #44128

### CODE REVIEW

File: src/gui/qgsmapcanvas.cpp [hunk 1]
```diff
@@ -205,17 +206,30 @@ QgsMapCanvas::QgsMapCanvas( QWidget *parent )
...
+    {
+      disconnect( mScreenDpiChangedConnection );
+      mScreenDpiChangedConnection = connect( window()->windowHandle()->screen(), &QScreen::physicalDotsPerInchChanged, this, [ = ]( qreal )
+      {
+        updateDevicePixelFromScreen();
+      } );
+      updateDevicePixelFromScreen();
+    } );
+
+    mScreenDpiChangedConnection = connect( window()->windowHandle()->screen(), &QScreen::physicalDotsPerInchChanged, this, [ = ]( qreal )
```
Comment (3nids) [inline src/gui/qgsmapcanvas.cpp:229]: This could use a simple connection rather than a lambda slot, no?

**Review by 3nids — APPROVED** [review]
Nice

File: src/gui/qgsmapcanvas.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/gui/qgsmapcanvas.cpp]: Yep, that's a leftover from an earlier approach

File: src/ui/qgsoptionsbase.ui [hunk 2]
```diff
@@ -332,7 +332,7 @@
        <item>
         <widget class="QStackedWidget" name="mOptionsStackedWidget">
          <property name="currentIndex">
-          <number>0</number>
+          <number>6</number>
```
Comment (domi4484) [inline src/ui/qgsoptionsbase.ui:335]: Would this change the default options view?

**Review by domi4484 — APPROVED** [review]

File: src/ui/qgsoptionsbase.ui (hunk 2 above)
Comment (m-kuhn) [reply src/ui/qgsoptionsbase.ui:335]: The default page is stored in the settings, so it shouldn't change, but we can keep it.

```suggestion
          <number>0</number>
```

**Comment by Houska1** [comment]
At @m-kuhn's request I tested the current patch to see if it solves https://github.com/qgis/QGIS/issues/41248.
Unfortunately not yet.

I've put more details there, but I think the solution needs to be a 3 way option called something like Map canvas DPI: 
1. Respect system UI scaling (default; uses `.logicalDpiX`; matches previous behaviour and font sizes match UI panels etc)
2. Use physical screen DPI (uses `.physicalDpiX`)
3. Scaling off = 96 DPI. (note this may be previous behaviour on linux but is not on Windows)

**Comment by m-kuhn** [comment]
> Unfortunately not yet.

Does it change anything at all?

**Comment by Houska1** [comment]
> > Unfortunately not yet.
> 
> Does it change anything at all?

I put all the details in https://github.com/qgis/QGIS/issues/41248

Before the patch, canvas `.outputDpi` used `.logicalDpiX` which was 120, which made symbols a bit too small. It should use `.physicalDpiX` to make them accurate size. With the patch, and the new option off (default), it's now 96 which is worse. With the option back on, it merely goes back to 120, which is what past behaviour was. I think you started from the premise that before the patch, QGIS used a fixed value of 96, which may be true in linux (I don't know) but it seems in Windows, QGIS/Qt used whatever `logicalDpiX` was.

**Comment by Houska1** [comment]
Thinking about this further. Would it make sense to allow the user to manually set a preferred (canvas) screen DPI? If undefined/zero use logical dpi as fallback (as you have done), but if they care enough to enter a number, use whatever they say (after restart)? I just fear that between Windows/Linux/Mac, internal/external monitors, Qt acting as go-between etc. whatever logic to autodetect it may fail in some edge cases. Reading https://doc.qt.io/qt-5/highdpi.html makes my head spin!

**Comment by m-kuhn** [comment]
What I would like to achieve in here is a configuration that represents 10 mm as 10 mm (physical dpi) in regular scenarios. So my notebook monitor and external monitor / my monitor, my coworkers monitor / my monitor, my mobile app show the same. This is something we currently fail to do.
On top of that we already now have the magnifier available.

We could certainly add more configuration options to allow for every attached output device a specific manually defined dpi configuration, but it will inherently add complexity which I would like to keep separate.

**Comment by Houska1** [comment]
> What I would like to achieve in here is a configuration that represents 10 mm as 10 mm (physical dpi) in regular scenarios. So my notebook monitor and external monitor / my monitor, my coworkers monitor / my monitor, my mobile app show the same. This is something we currently fail to do.

Understood and agreed. I guess I'm just more pessimistic that Qt gets it reliably right, across all possible platform/configurations. 

The version I tested this morning went to 96 with the option off. I think your "use logical dpi as fallback" change fixes that. But with the option on, it was using 120 (=logical Dpi) and not actually picking up on my physical Dpi of 158. I don't think that's fixed yet, though I'll poke a bit more to make sure.

> We could certainly add more configuration options to allow for every attached output device a specific manually defined dpi configuration, but it will inherently add complexity which I would like to keep separate.

Agree that a separate setting per output device would be too much complexity. I was thinking a single manually set UI option used by QGIS until changed by the user. A bit inelegant if they change screens often, but as a workable solution if getting the right physical Dpi is a challenge. 

I haven't tried, but I'm just not convinced that -- say -- plugging in an older-generation 1024x768 external monitor via a VGA cable to a video card is going to reliably generate a different physical DPI setting in the operating system based on whether it's a 17", 19", or 21" diagonal monitor. 

I'm poking around and seeing there's something called EDID whereby VGA, HDMI, and DVI-X should all allow a monitor to pass its size to the graphics card and hence to the operating system, but the last 2 answers at https://superuser.com/questions/1597994/does-windows-know-physical-size-of-external-monitor don't inspire confidence that one can depend on it.

**Comment by Houska1** [comment]
I've reconfirmed and I do have `Respect screen DPI on`, and have restarted. And yet my canvas (from visual inspection as well as from `iface.mapCanvas().mapSettings().outputDpi()` in the Python console) is 120, which is the logical DPI, while `iface.mainWindow().physicalDpiX()` is 158. 

What I don't understand is why that's happening, since I see your code in QgsMapCanvas::updateDevicePixelFromScreen() does refer to physical DPI!

**Comment by m-kuhn** [comment]
@Houska1 thanks for the in depth testing. Can you give the latest one another try? It seems it was related to a missing window at creation time, so initialisation can only be done after the showEvent

**Comment by Houska1** [comment]
Hurrah! Now works for me as intended on Windows, both on my laptop screen (phys DPI 158) and a random external monitor (phys DPI 94). And adjusts if I move the QGIS window from one to the next.

**Comment by m-kuhn** [comment]
Thanks for testing and the ideas.
Let's see how this performs in "real world" (e.g. checking out a beamer would be nice) and if it is better suited as default than the logical dpi.

**Comment by nirvn** [comment]
@m-kuhn , good job tackling this. A nice step forward.

**Pull request merged**